training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py

```diff
@@ -0,0 +1,291 @@
+    no_decay = ["bias", "LayerNorm.weight", "layer_norm.weight"]
+    optimizer_grouped_parameters = [
+        {
+            "params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
+            "weight_decay": args.weight_decay,
+        },
+        {
+            "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+            "weight_decay": 0.0,
+        },
+    ] 
+
+    optimizer = torch.optim.AdamW(optimizer_grouped_parameters, lr=args.learning_rate)
+
+    # Scheduler and math around the number of training steps.
+    overrode_max_train_steps = False
+    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
+    if args.rank==0:
+        print(f"Number of update steps per epoch {num_update_steps_per_epoch}")
+    if args.max_train_steps is None:
+        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
+        overrode_max_train_steps = True
+
+    lr_scheduler = get_scheduler(
+        name=args.lr_scheduler_type,
+        optimizer=optimizer,
+        num_warmup_steps=args.num_warmup_steps * args.gradient_accumulation_steps,
+        num_training_steps=args.max_train_steps * args.gradient_accumulation_steps,
+    )
+
+    start = time.time()
+    device = torch.device(f"cuda:{args.local_rank}")
+
+    for epoch in range(args.num_train_epochs):
+
+        model.train()
+        total_steps=0
+        fsdp_loss = torch.zeros(2).to(args.local_rank)
+
+        for _, batch in enumerate(tqdm(train_dataloader,disable=not (args.rank==0))):
+
+            batch = {k: v.to(device) for k, v in batch.items()}
+            output = model(**batch)
+            loss = output["loss"]
+            loss.backward()
+            fsdp_loss[0] += loss.item()
+            fsdp_loss[1] += len(batch["input_ids"])
+        
+            optimizer.step()
+            lr_scheduler.step()
+            optimizer.zero_grad()
+            total_steps += 1
+            if args.max_steps is not None and total_steps > args.max_steps:
+                break
+             
+
+        torch.distributed.all_reduce(fsdp_loss, op=torch.distributed.ReduceOp.SUM)
+        train_loss = fsdp_loss[0] / fsdp_loss[1]
+        train_ppl = torch.exp(train_loss)
+
+        if args.rank==0:
+            print(f"******{epoch=}: {train_ppl=} {train_loss=}******")
+        
+
+        model.eval()
+        eval_loss = 0
+        fsdp_eval_loss = torch.zeros(2).to(args.local_rank)
+        for steps, batch in enumerate(tqdm(eval_dataloader,disable=not (args.rank==0))):
+            batch = {k: v.to(device) for k, v in batch.items()}
+            with torch.no_grad():
+                outputs = model(**batch)
+            loss = outputs["loss"]
+
+            fsdp_eval_loss[0] += loss.item()
+            fsdp_eval_loss[1] += len(batch["input_ids"])
+            if args.max_steps is not None and steps > args.max_steps:
```

Comment:
Do we need this check for `eval`? I think it can be removed.

Reply:
If the `max_steps` argument is set, then we don't want the number of validation steps (i.e. the number of batches fed to model for validation) to exceed it as well.